core/src/agent/actions/add_link.rs

```diff
@@ -0,0 +1,158 @@
+use action::{Action, ActionWrapper};
+use agent::state::{ActionResponse, AgentState};
+use chain::SourceChain;
+use context::Context;
+use hash_table::{
+    links_entry::{LinkActionKind, LinkEntry},
+    sys_entry::ToEntry,
+    HashTable,
+};
+use instance::Observer;
+use std::sync::{mpsc::Sender, Arc};
+
+/// Do the AddLink Action against an agent state:
+/// 1. Validate Link
+/// 2. Commit LinkEntry
+/// 3. Add Link metadata in HashTable
+pub(crate) fn reduce_add_link(
+    _context: Arc<Context>,
+    state: &mut AgentState,
+    action_wrapper: &ActionWrapper,
+    _action_channel: &Sender<ActionWrapper>,
+    _observer_channel: &Sender<Observer>,
+) {
+    let action = action_wrapper.action();
+    let link = unwrap_to!(action => Action::AddLink);
+
+    // TODO #277
+    // Validate Link Here
+
+    // Create and Commit a LinkEntry on source chain
+    let link_entry = LinkEntry::from_link(LinkActionKind::ADD, link);
+    let res = state.chain().clone().push_entry(&link_entry.to_entry());
+    let mut response = if res.is_ok() {
+        Ok(res.unwrap().entry().clone())
+    } else {
+        Err(res.err().unwrap())
+    };
+
+    // Add Link to HashTable (adds to the LinkListEntry Meta)
+    let res = state.chain().table().add_link(link);
+    if res.is_err() {
+        response = Err(res.err().unwrap());
+    }
+
+    // Insert reponse in state
+    state.actions().insert(
+        action_wrapper.clone(),
+        ActionResponse::LinkEntries(response),
```

Comment:
Do you think it would be benficial to use an and_then function here? I think it has the potential to eliminate some code

Reply:
Awesome! Looks good to me. Will approve